opacclient/libopac/src/main/java/de/geeksfactory/opacclient/apis/Wise.kt

```diff
@@ -0,0 +1,563 @@
+package de.geeksfactory.opacclient.apis
+
+import de.geeksfactory.opacclient.i18n.StringProvider
+import de.geeksfactory.opacclient.networking.HttpClientFactory
+import de.geeksfactory.opacclient.networking.NotReachableException
+import de.geeksfactory.opacclient.objects.*
+import de.geeksfactory.opacclient.searchfields.BarcodeSearchField
+import de.geeksfactory.opacclient.searchfields.DropdownSearchField
+import de.geeksfactory.opacclient.searchfields.SearchField
+import de.geeksfactory.opacclient.searchfields.SearchQuery
+import de.geeksfactory.opacclient.utils.jsonObject
+import okhttp3.Interceptor
+import okhttp3.MediaType.Companion.toMediaTypeOrNull
+import okhttp3.RequestBody.Companion.toRequestBody
+import okhttp3.Response
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+import java.util.*
+import javax.crypto.Mac
+import javax.crypto.spec.SecretKeySpec
+import kotlin.experimental.and
+import kotlin.math.floor
+
+data class Authentication(val libraryId: String, val patronSystemId: String,val token: String, val membershipExpiryDate: org.joda.time.LocalDate )
+
+lateinit var apiKeyId: String
+lateinit var apiKey: String
+lateinit var applicationName: String
+var auth: Authentication? =  null
+
+/**
+ * OpacApi implementation for Wise owned by OCLC https://www.oclc.org/en/wise.html
+ *
+ *
+ * @author Henk Klijn Hesselink, April 2021
+ */
+
+
+class WiseKeyInterceptor : Interceptor {
+    override fun intercept(chain: Interceptor.Chain): Response {
+
+        val originalRequest = chain.request()
+        val requestBuilder = originalRequest.newBuilder()
+                .header("WISE_KEY", genWisekey())
+        if ( auth != null) requestBuilder.addHeader("Authorization", auth!!.token)
+        val request = requestBuilder.build()
+        return chain.proceed(request)
+    }
+
+    fun genWisekey(): String {
+        val epochDay: String  = floor(Date().time / 8.64e7).toLong().toString()
+        val data = epochDay + applicationName
+
+        val mac = Mac.getInstance("HmacSHA256")
+        val secret_key = SecretKeySpec(apiKey.toByteArray(charset("UTF-8")), "HmacSHA256")
+        mac.init(secret_key)
+        val signature= bytesToHex(mac.doFinal(data.toByteArray(charset("UTF-8"))))
+        return "$apiKeyId:$signature"
```

Comment:
Is there any documentation on this algorithm that we could link to here for future reference?

Reply:
No there is no official public documentation on this algorithm. it can be found though hidden in js files op Wise web OPAC. All official 3rd party vendors which needs to consumes endpoint of the RESTapi do get a copy from OCLC. OPAC  App is not on that list (yet)